firedrake/assemble.py

```diff
@@ -480,17 +480,25 @@ def all_integer_subdomain_ids(self):
 
     @cached_property
     def global_kernels(self):
-        return tuple(_make_global_kernel(self._form, tsfc_knl, self.all_integer_subdomain_ids,
-                                         diagonal=self.diagonal,
-                                         unroll=self.needs_unrolling(tsfc_knl, self._bcs))
-                     for tsfc_knl in self.local_kernels)
+        return tuple(
+            _make_global_kernel(
+                self._form, tsfc_knl, subdomain_id, self.all_integer_subdomain_ids,
+                diagonal=self.diagonal, unroll=self.needs_unrolling(tsfc_knl, self._bcs)
+            )
+            for tsfc_knl in self.local_kernels
+            for subdomain_id in tsfc_knl.kinfo.subdomain_id
```

Comment:
Yes that would launch 4 kernels. Your suggestion is clearly the right thing to do. Thanks!

Reply:
Actually I believe this is the right thing to do. The example given in the [UFL PR](https://github.com/FEniCS/ufl/pull/92#issue-1149360544) is:

```py
a = inner(u, v)*dx((1, 2, 3)) + inner(grad(u), grad(v))*dx(2) + inner(u, v)*dx
```

which yields 2 `IntegralData` objects:

> On this branch, we get two IntegralData, one for inner(u,v)*dx((1,3,"everywhere")) and one for (inner(u,v) + inner(grad(u), grad(v)))*dx((2,)).

If we only executed a kernel over the union of `(1, 3, "everywhere")` then we would only visit points in subdomains `1` and `3` once, rather than twice as expected.

Reply:
Hmm, I see. So UFL knows nothing about the decomposition of the subdomains (how can it?) and so even though you get two kernels, the first one needs to be executed for each subdomain in turn.

One way you could consider changing that is by relaxing an `op2.Subset` into a [multiset](https://en.wikipedia.org/wiki/Multiset) (would only need API, not codegen changes I think) so that you could say:
```
subset = op2.Subset(mesh.cell_set, [1, 1, 1])
```
To integrate over element `1` three times.

Reply:
Possibly. Though this may actually cause codegen issues if one of the subdomains were `"everywhere"` since we wouldn't then want to pass in the subset indices to the global kernel.

Reply:
It (at least used to be) an error if a set of forms coming out of `compute_form_data` had `"everywhere"` as their subdomain id.